ethpm/contract.py

```diff
@@ -0,0 +1,119 @@
+            raise BytecodeLinkingError("Contract factory has no linkable bytecode.")
+        if self.is_bytecode_linked:
+            raise BytecodeLinkingError(
+                "Bytecode for this contract facotry has already been linked."
+            )
+        self.validate_attr_dict(self, attr_dict)
+        self._link_links(self, attr_dict)
+
+    def _link_links(self, attr_dict: Dict[str, str]) -> None:
+        # Fill `bytecode` link references
+        if self.deployment_link_refs:
+            link_fns = (
+                apply_link_reference(offset, ref["length"], attr_dict[ref["name"]])
+                for ref in self.deployment_link_refs
+                for offset in ref["offsets"]
+            )
+            linked_bytecode = cytoolz.pipe(self.bytecode, *link_fns)
+            self.bytecode = linked_bytecode
+        # Fill `bytecode_runtime` link references
+        elif self.runtime_link_refs:
+            link_fns = (
+                apply_link_reference(offset, ref["length"], attr_dict[ref["name"]])
+                for ref in self.runtime_link_refs
+                for offset in ref["offsets"]
+            )
+            linked_bytecode_runtime = cytoolz.pipe(self.bytecode_runtime, *link_fns)
+            self.bytecode_runtime = linked_bytecode_runtime
+
+        self.is_bytecode_linked = True
+
+    def validate_attr_dict(self, attr_dict: Dict[str, str]) -> None:
+        """
+        Validates that ContractType keys in attr_dict reference existing manifest ContractTypes.
+        """
+        attr_dict_names = list(attr_dict.keys())
+        all_link_refs = (
+            # Ignored type in case one of the following is None
+            self.deployment_link_refs  # type: ignore
+            + self.runtime_link_refs  # type: ignore
+        )  # type: List[Dict[str, Any]]
+        all_link_names = [ref["name"] for ref in all_link_refs]
+        if set(attr_dict_names) != set(all_link_names):
+            raise BytecodeLinkingError(
+                """All link references must be defined when calling """
+                """`link_bytecode` on a contract factory."""
+            )
+        for address in list(attr_dict.values()):
+            if not is_address(address):
+                raise BytecodeLinkingError(
+                    "Address: {0} as specified in the attr_dict is not a valid address.".format(
+                        address
+                    )
+                )
+
+
+@cytoolz.curry
+def apply_link_reference(
+    offset: int, length: int, value: str, bytecode: bytes
+) -> bytes:
+    """
+    Returns the new bytecode with `value` put into the location indicated by `offset` and `length`.
+    """
+    try:
+        validate_empty_bytes(offset, length, bytecode)
+    except ValidationError:
+        raise BytecodeLinkingError("Link references cannot be applied to bytecode")
+
+    new_bytes = (
+        # Ignore linting error b/c black conflict
```

Comment:
Probably not, formatting rules for whitespace around `:` are different b/w black and flake8. since black is super opinionated, i'm just ignoring the flake8 complaint here. Unless it's better to update `tox.ini`?